src/lib/structures/uid.ml

```diff
@@ -0,0 +1,75 @@
+(*                                                                        *)
+(*     Alt-Ergo: The SMT Solver For Software Verification                 *)
+(*     Copyright (C) 2013-2024 --- OCamlPro SAS                           *)
+(*                                                                        *)
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module DStd = Dolmen.Std
+module DE = DStd.Expr
+
+type t =
+  | Hstring : Hstring.t -> t
+  | Dolmen : 'a DE.id -> t
+
+let[@inline always] of_dolmen id = Dolmen id
+let[@inline always] of_hstring hs = Hstring hs
+let[@inline always] of_string s = of_hstring @@ Hstring.make s
+
+let hash = function
+  | Hstring hs -> Hstring.hash hs
+  | Dolmen { index; _ } -> (index :> int)
+
+let pp ppf = function
+  | Hstring hs -> Hstring.print ppf hs
+  | Dolmen id -> DE.Id.print ppf id
+
+let show = Fmt.to_to_string pp
+
+let equal u1 u2 =
+  match u1, u2 with
+  | Hstring hs1, Hstring hs2 -> Hstring.equal hs1 hs2
+  | Dolmen id1, Dolmen id2 -> DE.Id.equal id1 id2
+  | _ ->
+    Hstring.equal (Hstring.make @@ show u1) (Hstring.make @@ show u2)
```

Comment:
Isn't this a complicated way of writing `String.equal (show u1) (show u2)`? The complexity should be the same since `Hstring.make` requires a linear scan.

Reply:
I thought it was a little bit faster because we compare integers if these `Hstring` already exist but we have to compute the hash of the strings, which may be linear in the length of them.